clang/lib/Parse/ParseStmt.cpp

```diff
@@ -2427,3 +2447,75 @@ bool Parser::ParseOpenCLUnrollHintAttribute(ParsedAttributes &Attrs) {
   }
   return true;
 }
+
+WhereClauseFact *Parser::ParseWhereClauseFact() {
+  SourceLocation Loc = Tok.getLocation();
+
+  // TODO: Handle bounds expression surrounded by parentheses, like:
+  // _Where ((((p : bounds(p, p + 1))))).
+  // Equality expressions surrounded by parentheses are already handled by
+  // invoking IgnoreValuePreservingCasts in ActOnEqualityOpFact.
+
+  if (Tok.is(tok::identifier) && NextToken().is(tok::colon)) {
+    IdentifierInfo *VarName = Tok.getIdentifierInfo();
+
+    // Consume the identifier.
+    ConsumeToken();
+    // Consume the ':' token.
+    ConsumeToken();
+
+    // Parse bounds expression.
+    ExprResult BoundsRes(ParseBoundsExpression());
```

Comment:
I am surprised by the error messages in the test cases where you get "expected expression" as an error message.  I would hope to get a "expected a bounds expression".   Perhaps this is a case of the test case containing only part of the error message.